test/conformance/protocol_test.go

```diff
@@ -0,0 +1,196 @@
+
+	resp, err := pkgTest.WaitForEndpointState(
+		pt.clients.KubeClient, pt.logger, domain,
+		pkgTest.Retrying(
+			func(resp *spoof.Response) (bool, error) {
+				if resp.StatusCode == http.StatusOK {
+					return true, nil
+				}
+
+				return true, fmt.Errorf("unexpected status: %d", resp.StatusCode)
+			},
+			http.StatusNotFound,
+		),
+		pt.t.Name(), test.ServingFlags.ResolvableDomain,
+	)
+	if err != nil {
+		pt.t.Fatalf("Failed to get a successful request from %s: %v", domain, err)
+	}
+
+	pt.logger.Infof("Got response: %s", resp.Body)
+
+	return resp
+}
+
+func (pt *protocolsTest) createService(options *test.Options) *v1alpha1.Service {
+	pt.logger.Infof("Creating service %q with options: %#v", pt.names.Service, options)
+
+	service := test.LatestService(test.ServingNamespace, pt.names, options)
+
+	svc, err := pt.clients.ServingClient.Services.Create(service)
+	if err != nil {
+		pt.t.Fatalf("Error creating service: %s", err.Error())
+	}
+
+	if err := test.WaitForServiceState(pt.clients.ServingClient, pt.names.Service, test.IsServiceReady, "ServiceIsReady"); err != nil {
+		pt.t.Fatalf("Service %s not marked Ready: %v", pt.names.Service, err)
+	}
+
+	pt.logger.Infof("Service %q created", pt.names.Service)
+
+	return svc
+}
+
+func (pt *protocolsTest) getDomain(service *v1alpha1.Service) string {
+	pt.names.Route = serviceresourcenames.Route(service)
+
+	pt.logger.Infof("Fetching domain from route %q", pt.names.Route)
+
+	if err := test.WaitForRouteState(pt.clients.ServingClient, pt.names.Route, test.IsRouteReady, "RouteIsReady"); err != nil {
+		pt.t.Fatalf("Route %s not marked Ready: %v", pt.names.Route, err)
+	}
+
+	route, err := pt.clients.ServingClient.Routes.Get(pt.names.Route, metav1.GetOptions{})
+	if err != nil {
+		pt.t.Fatalf("Error fetching Route %s: %v", pt.names.Route, err)
+	}
+
+	pt.logger.Infof("Got domain %q", route.Status.Domain)
+
+	return route.Status.Domain
+}
+
+type protocol struct {
+	Major int `json:"protoMajor"`
+	Minor int `json:"protoMinor"`
+}
+
+func (p *protocol) String() string {
+	return fmt.Sprintf("HTTP/%d.%d", p.Major, p.Minor)
+}
+
+func portOption(portname string) *test.Options {
+	options := &test.Options{}
+
+	if portname != "" {
+		options.ContainerPorts = []corev1.ContainerPort{{Name: portname, ContainerPort: 8080}}
+	}
+
+	return options
+}
+
+func TestProtocols(t *testing.T) {
+	tests := []struct {
+		Name     string
+		PortName string
+		Want     protocol
+	}{{
+		Name:     "HTTP/2.0",
+		PortName: "h2c",
+		Want:     protocol{Major: 2, Minor: 0},
+	}, {
+		Name:     "HTTP/1.1",
+		PortName: "http1",
+		Want:     protocol{Major: 1, Minor: 1},
+	}, {
+		Name:     "Default",
+		PortName: "",
+		Want:     protocol{Major: 1, Minor: 1},
+	}}
+
+	for _, tt := range tests {
+		t.Run(tt.Name, func(t *testing.T) {
+			var pt protocolsTest
+
+			pt.setup(t)
+			defer pt.teardown()
+
+			options := portOption(tt.PortName)
+
+			service := pt.createService(options)
+			domain := pt.getDomain(service)
```

Comment:
Are those two lines (+ functions) the same as:

```golang
objects, err := test.CreateRunLatestServiceReady(logger, clients, &names, options)
if err != nil {
	t.Fatalf("Failed to create service %v", err)
}
domain := objects.Route.Status.Domain
```

?

Reply:
No, I think that only covers up to `pt.createService()`, since it only waits for the Service to be ready. 

From what I've seen, the Service can be ready before the Route is ready, and before the service is actually ready to serve requests. This is what `pt.getDomain` checks with `WaitForEndpointState`.

Reply:
The latter condition should be fixed by now IIRC.

Reply:
Wait, my mistake, `pt.getDomain` doesn't even `WaitForEndpointState`. Thanks for the suggestion, I'll switch to `test.CreateRunLatestServiceReady`.